### ISSUE TITLE
Option to require return statements for Optional return types

### ISSUE THREAD

**Issue by jtlapp** [microsoft/pyright#2184]
I would like a configuration option that requires functions to explicitly include a `return` when the return type can be non-`None`.

I just uncovered a bug in a function of mine that returns an `Optional` value. The bug was that a particular path through the function should have returned a non-`None` value but instead had no `return` statement at all, resulting in a return value of `None`. I sometimes organize functions as a series of conditional returns to prevent excessive indentation.

I'm a statically-typed-language kind of person and am trying to reap similar benefits from Python. As an aside, in case you have any suggestions, I am frequently comparing functions to values and not getting pyright errors (e.g. `if table.find_value == "x"`), when I intended to actually call the function (e.g. `if table.find_value() == "x"`).

Here's the configuration I use:

```
{
    "typeCheckingMode": "strict",
    "reportImportCycles": "none",
    "reportUninitializedInstanceVariable": "error",
}
```

**Comment by jakebailey** [comment]
TS has the equivalent for `undefined` returns: https://www.typescriptlang.org/tsconfig#noImplicitReturns

**Comment by erictraut** [comment]
That's an interesting idea. As Jake said, there is an option like that in TypeScript. I don't think we could turn on a feature like this by default, even in strict mode, because Python code often relies on the implicit return behavior. I think this is worth prototyping to see whether it uncovers any legitimate bugs in some largish Python code bases.

As for the other issue (unintended comparison to a function), that rule would be a bit more difficult to formulate. It would probably require us to hard-code a bunch of knowledge about many stdlib type semantics — something we're very reluctant to do. From a type perspective, the expression `table.find_value == "x"` is fine because the `__eq__` operator is typed to accept any `object` as an input parameter.

**Comment by jtlapp** [comment]
Thanks for looking into it.

I have an idea about how to deal with the function comparison issue, but I want to get back with you after I've thought more about it.

FYI, I think an update to Pylance just broke type checking in two places where I define variables with a `Callable` type. Other occurrences of such variables are working, so I'm going to have to put some time into figuring out whether to report a bug. This is code I haven't touched in weeks, so I figure it must be an update to pyright.

**Comment by jtlapp** [comment]
Regarding the update problem, best I can tell, import cycles that were not causing pyright trouble before are causing pyright trouble now. When I remove one of the import and its associated code, the type errors go away.

**Comment by erictraut** [comment]
Import cycles are pernicious in Python because they have stateful (and therefore order-dependent) side effects, yet these side effects are almost impossible for tooling (or humans) to reason about. In my experience, import cycles can result in subtle bugs that manifest when seemingly-unrelated changes are made to the code.

I can't think of any specific changes in pyright that would cause a break here, but it's possible that type resolution order changed slightly in some subtle way. If that's the case, I wouldn't consider that a bug in the type evaluator but rather a fragility in the code being modeled. If you are able to find a self-contained repro, I'd be happy to take a look at it.

Pyright does offer a diagnostic check for import cycles (`reportImportCycles`). It's not enabled by default, but you might want to turn it on.

**Comment by jtlapp** [comment]
I checked out weeks-old code, and that code has the same new type errors, so it's definitely newly introduced.

I've been programming for 40 years, 22 in Java, and I've never faced this sort of compiler confusion before. It must be the single-pass compilation. I did deal with some cyclic issues in JavaScript and TypeScript, but never like this.

I'll factor the circular dependency out. Thanks for your help!

**Comment by erictraut** [comment]
I prototyped this feature. In addition to exempting functions with a declared `None` return type, I also needed to exempt:
1. Functions with no explicit declared return type
2. Function declarations in stub files
3. Generator functions
4. Functions with an explicit `Any` return type
5. Abstract methods

I ran it on a large Python code base (~250K lines) that is heavily typed, most of it passing in "strict" mode. The new diagnostic check flagged 8 instances, and all of them were false positives (not real bugs).

I'll discuss further with my colleagues, but based on these results, I don't know if this feature is useful enough to justify adding another configuration option and added complexity to the type checker.
